### PR TITLE
Use macros to define default copy constructor and assignment operator

### PR DESCRIPTION
**Pull request by MaartenBent**
A different solution for `= default`, see discussion here: https://github.com/wxWidgets/wxWidgets/commit/413fdfbfa47409af822021036b01f950ba3b2406#r38444880

### CODE REVIEW

File: include/wx/defs.h [hunk 1]
```diff
@@ -3042,6 +3039,26 @@ typedef const void* WXWidget;
 #define DECLARE_NO_ASSIGN_CLASS(classname) \
     wxDECLARE_NO_ASSIGN_CLASS(classname);
 
+/*  --------------------------------------------------------------------------- */
+/*  macros to define a class with default copy constructor and assignment operator
+/*  --------------------------------------------------------------------------- */
```
Comment (vadz) [inline include/wx/defs.h:3044]: This comment results in a ton of warning in Travis builds, it should be rewritten to either use single `/* ... */` around it or add missing `*/` to the second line.

File: interface/wx/defs.h [hunk 2]
```diff
@@ -1561,6 +1561,24 @@ typedef double wxDouble;
  */
 #define wxDECLARE_NO_COPY_TEMPLATE_CLASS_2(classname, arg1, arg2)
 
+/**
+    This macro can be used in a class declaration to enable declaring the
+    default copy constructor and assignment operator.
+ */
```
Comment (vadz) [inline interface/wx/defs.h:1567]: Could you please add `@since 3.1.4` (and also below)?

**Review by vadz — CHANGES_REQUESTED** [review]
Sorry but I really preferred my name here as IMO `wxDECLARE_DEFAULT_CLASS()` is unclear -- default what? Personally I'd assume it means the default ctor and dtor are default too, but this macro doesn't do this.

IMHO `wxDECLARE_DEFAULT_COPY_CLASS()` for both members is fine. `wxDECLARE_DEFAULT_ASSIGN_CLASS()` would be consistent with `wxDECLARE_NO_ASSIGN_CLASS()` so I'm not really against it, but I just don't see when would it be useful: I can't think of a class for which default copy ctor wouldn't work, but default assignment operator would.

So I'd keep things simple and have just one `wxDECLARE_DEFAULT_COPY_CLASS()`.

**Comment by MaartenBent** [comment]
The problem is that using `wxDECLARE_DEFAULT_COPY_CLASS` for both members will not work for `wxUniCharRef` because it already has a `operator=`:

```
error C2535: 'wxUniCharRef &wxUniCharRef::operator =(const wxUniCharRef &)': member function already defined or declared
```

 That is why I split it up into multiple macros.
`wxUniCharRef` would still need a separate macro with only the copy constructor, but with what name?

Maybe I should just revert it and only keep `wxHAS_MEMBER_DEFAULT`.

**Comment by vadz** [comment]
Ah, I see. This case (unlike the one for `DEFAULT_ASSIGN`) is indeed possible, but I hadn't realized we already had it.

We could have `wxDECLARE_DEFAULT_COPY_CTOR()` for it? This, at least, should be absolutely clear.

File: include/wx/animate.h [hunk 3]
```diff
@@ -38,10 +38,7 @@ class WXDLLIMPEXP_CORE wxAnimation : public wxObject
     wxAnimation();
     explicit wxAnimation(const wxString &name, wxAnimationType type = wxANIMATION_TYPE_ANY);
 
-#ifdef wxHAS_MEMBER_DEFAULT
-    wxAnimation(const wxAnimation&) wxMEMBER_DEFAULT;
-    wxAnimation& operator=(const wxAnimation&) wxMEMBER_DEFAULT;
-#endif
+    wxDECLARE_DEFAULT_COPY_CLASS(wxAnimation);
```
Comment (paulcor) [inline include/wx/animate.h:41]: If member-default is not supported, this can cause a warning:
`warning: extra ‘;’ [-Wpedantic]`

**Review by paulcor — COMMENTED** [review]
As usual, I now regret opening my mouth. Just for the record, I won't be using `wxDECLARE_DEFAULT_COPY_CLASS()`, as it adds no value over just a simple, clear comment.

File: include/wx/animate.h (hunk 3 above)
Comment (MaartenBent) [reply include/wx/animate.h:41]: Yes, I can trigger it too: `warning: extra ';' after member function definition [-Wextra-semi]`.

**Comment by MaartenBent** [comment]
I'll revert the changes for now. I can't think of a way to silence the `-Wextra-semi` warning. Maybe the macros could be added once there is a need for them.

**Comment by vadz** [comment]
> I'll revert the changes for now. I can't think of a way to silence the `-Wextra-semi` warning.

Well, the standard way is to make the macro expand to `struct wxDummyDeclToEatSemicolon` when it's "not defined", but I'm not sure if it's worth it at this point.

**Pull request merged**